types/inferschematype.d.ts

```diff
@@ -0,0 +1,61 @@
+import { Schema, InferSchemaType, SchemaType, SchemaTypeOptions } from 'mongoose';
+
+declare module 'mongoose' {
+  type ObtainDocumentType<DocDefinition, DocType = any> =
+    DoesDocTypeExist<DocType> extends true ? DocType : {
+      [K in keyof (RequiredPaths<DocDefinition> &
+        OptionalPaths<DocDefinition>)]: ObtainDocumentPathType<DocDefinition[K]>;
+    };
+
+  type InferSchemaType<SchemaType> = SchemaType extends Schema<infer DocType, any, any, any, infer DocDefinition>
+    ? DoesDocTypeExist<DocType> extends true ? DocType : DocDefinition
+    : unknown;
+
+  type ObtainSchemaGeneric<TSchema, name extends 'DocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'DocDefinition' | 'StaticsMethods'> =
+    TSchema extends Schema<infer DocType, infer M, infer TInstanceMethods, infer TQueryHelpers, infer DocDefinition, infer StaticsMethods>
+    ? { DocType: DocType, M: M, TInstanceMethods: TInstanceMethods, TQueryHelpers: TQueryHelpers, DocDefinition: DocDefinition, StaticsMethods: StaticsMethods }[name]
+    : never;
+}
+
+type DoesDocTypeExist<DocType> = keyof DocType extends string ? true : false;
+
+type RequiredPathBaseType = { required: true | [true, string | undefined] }
+
+type PathWithTypePropertyBaseType = { type: any }
```

Comment:
This unfortunately won't work with custom `typeKey`, see discussion from #11482. We can consider just not supporting `typeKey` in TypeScript, in which case this would be fine.

Reply:
This feature is support typeKey option property :)

Reply:
I've fixed the intellisense issue when providing custom typeKey Can I push it in this feature or will create new PR ?

Reply:
![Untitled](https://user-images.githubusercontent.com/50488054/158070963-ea4e76d2-d12a-4ecb-bb27-d7a2f9b84ecb.png)

Reply:
I've refacted some other types to show custom typeKey as a suggestion, should I create new PR for that after merging this one OR Is it okey to have it with in this PR ?
![Untitled](https://user-images.githubusercontent.com/50488054/158073894-cab9b5e3-1178-41dd-8594-6897d76f7c39.png)